neo4j: resolve compgen dependency error

DavidRConnell opened the pull request:
- Fix #130112.
- neo4j was unable to find `compgen` at runtime.
- Need to use `bashInteractive` version of bash since default bash does
not have `compgen`.
- Add a compgen wrapper shell script to pass `compgen` calls to `bash
-c 'compgen "$@"'` in order to run `compgen` outside of a `bash`
shell.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Fixes issue #130112 where neo4j cannot find compgen.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
neo4j: resolve compgen dependency error

- Fix #130112.
- neo4j was unable to find `compgen` at runtime.
- Need to use `bashInteractive` version of bash since default bash does
not have `compgen`.
- Add a compgen wrapper shell script to pass `compgen` calls to `bash
-c 'compgen "$@"'` in order to run `compgen` outside of a `bash`
shell.

neo4j: remove compgen from global environment

r-rmcgibbo commented:
Result of `nixpkgs-review pr 130857` at b3b2fa5e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>neo4j</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/servers/nosql/neo4j/default.nix:16:3:
```
   |
16 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/servers/nosql/neo4j/default.nix:39:5:
```
   |
39 |     license = licenses.gpl3;
   |     ^
```
</li>
  </ul>
</details>

happysalada commented on pkgs/servers/nosql/neo4j/default.nix:
```diff
@@ -11,18 +11,24 @@ stdenv.mkDerivation rec {
     cp -R * "$out/share/neo4j"
 
     mkdir -p "$out/bin"
+    compgen_wrapper="$out/share/neo4j/bin/compgen"
```
small question.
the mkdir above create `$out/share/neo4j`
is the `bin` subdirectory created as well somehow ?
cat won't create the subdirectory, I think.

happysalada replied:
oh, everything gets copied from the $out in the step above. Nevermind then.

happysalada commented on pkgs/servers/nosql/neo4j/default.nix:
```diff
@@ -11,18 +11,24 @@ stdenv.mkDerivation rec {
 
     mkdir -p "$out/bin"
+    compgen_wrapper="$out/share/neo4j/bin/compgen"
+    cat << _EOF_ > $compgen_wrapper
```
it's really a nitpick, but I've never seen the file redirection after the EOF, is there a reason for this?
did you consider the following
```
cat > "$comgen_wrapper" << _EOF_
...
```

happysalada replied:
Actually looking around, I've now seen some more occurences of this, feel free to ignore if you feel strongly about that choice.

DavidRConnell replied:
I'm used to having file redirects being at the end of the line and didn't even think about it. But think your way is clearer. Made the change.

happysalada commented:
Result of `nixpkgs-review pr 130857` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>neo4j</li>
  </ul>
</details>

happysalada commented:
Just tested and it seems the fix is working fine
```
[nix-shell:~/.cache/nixpkgs-review/pr-130857/results/neo4j/bin]$ neo4j version
neo4j 3.5.14
```

I'm curious about your opinion on the file redirection after the HEREDOC . I'm fine to merge both ways.

happysalada commented:
(sorry it took so long for somebody to give you a review!)

happysalada commented:
I just noticed that this is using jre8 and that the repo recommends java11
This is more of a suggestion if you are interested, no need to address this in this PR or at all if you're not interested.

Commit:
neo4j: apply suggestions to conform with nixpkgs-hammering

Commit:
neo4j: bump to version 4.3.2

- Neo4j discontinued neo4j-import in version 4.
- Also updates java to 11 as specified by neo4j.

DavidRConnell commented:
If we're updating java might as well bump to version 4. If there's a reason not to go ahead with that I can go back and change the commit to just update java.

DavidRConnell commented:
There's also issues with [file permissions](https://neo4j.com/docs/operations-manual/current/configuration/file-locations/#file-locations-permissions). `neo4j` needs write access to a few of the directories installed to, the read-only, `/nix/store/` causing it to fail to run. The only work around I've managed is to either place them in `/var/lib/neo4j` (but that requires manually creating the directory since nix doesn't have permission to write to `/var` at install time) or hardcode my home directory into the conf file. I've asked in a few places but can't find a good solution. I'll create an issue for it but want to point out that `neo4j` still won't run as is.

SuperSandro2000 commented on pkgs/servers/nosql/neo4j/default.nix:
```diff
@@ -1,36 +1,47 @@
-{ lib, stdenv, fetchurl, makeWrapper, jre, which, gawk }:
+{ lib, stdenv, fetchurl, makeWrapper, jre, which, gawk, bashInteractive }:
 
 with lib;
 
```
```suggestion
```

SuperSandro2000 commented on pkgs/servers/nosql/neo4j/default.nix:
```diff
@@ -1,36 +1,47 @@
   };
 
-  nativeBuildInputs = [ makeWrapper ];
+  nativeBuildInputs = [ makeWrapper bashInteractive ];
```
Why does this need to be the interactive bash?

SuperSandro2000 commented on pkgs/servers/nosql/neo4j/default.nix:
```diff
@@ -1,36 +1,47 @@
 
   meta = with lib; {
     description = "A highly scalable, robust (fully ACID) native graph database";
     homepage = "http://www.neo4j.org/";
```
```suggestion
    homepage = "https://www.neo4j.org/";
```

happysalada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20003,7 +20003,7 @@ in
 
   neo4j = callPackage ../servers/nosql/neo4j {
-    jre = jre8_headless;
+    jre = openjdk11;
```
I'm not exactly sure if neo4j runs properly with headless or not.
If it can, using openjdk11_headless would reduce the size of the resulting closure.
If we are not sure, I would prefer to test with openjdk11_headless

DavidRConnell replied:
I tried running a few things using the headless and it seems to work fine so made the switch.

happysalada commented:
Thank you for your work!

Regarding the file permission issue. How are you trying to run neo4j ? Are you using the nixosModule or are you just trying to install the binary and run it from the cli ?
There is a nixosModule (under nixos/modules/services/databases/neo4j.nix), that explicitely configures the data directory and defines a "HOME". Is that not working as intended ? If not, feel free to post the errors, I'm happy to go through them with you.
If you have questions, feel free to post them here.

happysalada commented:
Somebody tried to make the 3 -> 4 upgrade but said it was non-trivial
https://github.com/NixOS/nixpkgs/pull/98443
Some modifications where made to the module which you can copy I think.
Feel free to post errors as you go through this.

If you don't feel like making the upgrade to 4 in this PR, that's fine too.

DavidRConnell commented:
Yeah, I had been running with CLI. I didn't realize how this package was connected to the module. I went with the neo4j package over the service because I wanted to have it as a dependency in a flake and I can't find anything on how to go about that. I'm also running neo4j on my computer rather than a remote server so I would rather start and stop it as needed than have it started on log in. (I need to understand services a little better so I'll figure that out later.)

As far as upgrading to 4 goes, I saw that comment and think they are refering to the migration process rather than the installation. Neo4j 4 doesn't gurentee that it will work with a db made with version 3, but it should be simple enough to recreate the db with version 4 so I don't know what the non-trivial parts mentioned were. Looks like there are a few changes to the conf files but nothing significant as far as I can tell.

Testing a module looks more involved than testing a package from a different branch of nixpkgs where I can run `nix run /path/to/repo#neo4j` or similar. I don't think I'm going to have time to ensure this works until next week or so. In the meantime we can hold off merging this or I can revert to version 3.

happysalada commented:
I'm happy for both.
Let me know whatever works best for you.

yeah testing a module you basically have to switch the nixpkgs running on your system to the one in the PR.
Otherwise run with the normal nixpkgs and make some overrides to coincide with the PR.

Since services are just system services, you can start and stop them at will. (I'm mentioning that in case you need to run on local and don't want to have it running all the time). You can also make some overrides to the module to not have your service start on boot. Manipulating nixosModules takes a bit of practice, but It worth it I think. Feel free to ask if anything doesn't make sense.

Artturin commented:
There's no longer a compgen call in `share/neo4j/bin/neo4j` or in any file
